micro/fivesandsixes_test.go

```diff
@@ -0,0 +1,80 @@
+package micro
+
+import (
+    "reflect"
+    "testing"
+
+	"github.com/awalterschulze/gominikanren/sexpr/ast"
+)
+
+var (
+    ast5 = ast.NewInt(5)
+    ast6 = ast.NewInt(6)
+)
+
+func fives(x *ast.SExpr) GoalFn {
+	return Zzz(DisjointO(
+		EqualO(
+			ast.NewVariable("x"),
+			ast.NewInt(5),
+		),
+		func() GoalFn {return fives(x)},
+	))()
+}
+
+func sixes(x *ast.SExpr) GoalFn {
+	return Zzz(DisjointO(
+		EqualO(
+			ast.NewVariable("x"),
+			ast.NewInt(6),
+		),
+		func() GoalFn {return sixes(x)},
+	))()
+}
+
+func TestFivesAndSixes(t *testing.T) {
+	// ((call/fresh (λ (q) (≡ q 5))) empty-state)
+	states := RunGoal(
+		1,
+		CallFresh(func(q *ast.SExpr) Goal {
+			return EqualO(
+				ast.NewVariable("q"),
+				ast.NewInt(5),
+			)
+		}),
+	)
+	got := Reify("q", states)
+    want := []*ast.SExpr{ast5}
+    if !reflect.DeepEqual(got, want) {
+        t.Fatalf("expected %#v but got %#v", got, want)
+    }
+
+	// (define (fives x) (disj (≡ x 5) (fives x)))
+	// ((call/fresh fives) empty-state)
+	states = RunGoal(
+		2,
+		CallFresh(func(x *ast.SExpr) Goal {
+			return func() GoalFn {return fives(x)}
+		}),
+	)
+	got = Reify("x", states)
+    want = []*ast.SExpr{ast5, ast5}
+    if !reflect.DeepEqual(got, want) {
+        t.Fatalf("expected %#v but got %#v", got, want)
+    }
+
+	states = RunGoal(
+		10,
+		CallFresh(func(x *ast.SExpr) Goal {
+			return DisjointO(
+				func() GoalFn {return fives(x)},
```

Comment:
Here you see the difficulty with lexical scope.
How does the ast parser deal with any of that, anyways?

Reply:
I think I forgot enough, that I need a bit more of an explanation and/or context, sorry.

Reply:
So in my first attempt at this, I wasnt passing x into fives at all (see the first testcase).
Which works in this case. In the general case though, we want to be able to compose defined functions.
Right now if you would name the var x in fives and y in sixes in that way, you wouldn't get the correct answer.
Solution for that is to pass the callfresh var into the fives func like I am doing here, but that needs the boilerplate you see here in order to work. So clearly that is not ideal.

I would like to be able to define fives in terms of x, and have that x be scoped only for the func, without having to pass the var into the function from the bigger func. That would need dealing with scope, which we currently aren't doing at all. This example is from the paper, I'm using it here to demonstrate the scope issue.
Did that make it more clear at all?

Reply:
Your pull request made finally made it clear to me.
https://github.com/awalterschulze/gominikanren/pull/9
Sorry I was a bit slow on understanding the issue, but thanks for fixing it <3